### PR TITLE
BccPagination

### PR DESCRIPTION
**Pull request by starkris51**
# Change summary
Added BccPagination

## Change type

-   [ ] No review
-   [ ] Big PR
-   [ ] Small PR
-   [ ] Refactor

**Closes #ISSUE_NUMBER** _or_ **Part of #ISSUE_NUMBER**

### CODE REVIEW

**Comment by starkris51** [comment]
@laurensgroeneveld Kunne du sjekke denne?

File: design-library/src/components/BccPagination/BccPagination.spec.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import { describe, it, expect } from "vitest";
+
+import { mount } from "@vue/test-utils";
+import BccPagination from "./BccPagination.vue";
+
+describe("BccPagination", () => {});
```
Comment (laurensgroeneveld) [inline design-library/src/components/BccPagination/BccPagination.spec.ts]: Can you add some tests for the pagination? Doesn't have to be too extensive, but I'm thinking some test for:

- It renders the right amount of page numbers
- It renders an ellipsis (on both sides) when needed
- Updating the rowsPerPage updates the rendered page numbers

File: design-library/src/tokens/tailwind/shadows.ts
Comment (laurensgroeneveld) [inline design-library/src/tokens/tailwind/shadows.ts:1]: I think this file was included by accident - we added some tools for rendering the shadows in #285 but are waiting on updated shadows I think before including them in the design library. In any case it would belong in a different PR.

File: design-library/src/components/BccPagination/BccPagination.css [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    .bcc-pagination-button-container button:hover {
+        @apply bg-secondary-hover text-interactive-hover
+    }
+
+    .bcc-pagination-button-selected {
+        @apply px-3.5 bg-secondary font-semibold text-interactive;
+    }
+ 
+    .bcc-pagination-button-unselected {
+        @apply px-3.5 bg-button-secondary text-secondary;
```
Comment (laurensgroeneveld) [inline design-library/src/components/BccPagination/BccPagination.css]: Even though it both comes down to white in the current theme, `*-button-*` styles should not be used outside buttons. Should be `bg-primary` instead.

File: design-library/src/components/BccPagination/BccPagination.css [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    .bcc-pagination-rowsperpage {
+        @apply w-20 text-caption
+    }
+
+    .bcc-pagination-button-container {
+        @apply inline-flex items-center
+    }
+
+
+    .bcc-pagination-button-container button {
```
Comment (laurensgroeneveld) [inline design-library/src/components/BccPagination/BccPagination.css]: I suggest making this a `bcc-pagination-button` base class instead to make it a bit clearer in the HTML. And to put the `unselected` styles on it by default, then just for a `selected` button an extra class `bcc-pagination-button-selected` is added. If someone wants to implement this on their own with this CSS then they don't have to struggle as much with swapping classes 😅.

File: design-library/src/components/BccPagination/BccPagination.vue
Comment (laurensgroeneveld) [inline design-library/src/components/BccPagination/BccPagination.vue:1]: It looks pretty good already, but I see some small differences from Figma.

Implemented version:
image

Figma version:
image

Differences I see here:
- The left arrow is not centered
- Each page number seems to have a double border at both sides instead of a single line (borders between numbers are not the same width as the top/bottom borders)
- Horizontal and vertical padding is a bit larger than in the design (should be 12px horizontal, 6px vertical) - this is part of the reason why the rendered one has a bit more height than the one from the design (33px vs 40px - although this one now has a fixed height - I would suggest to remove it).

- Additionally, the ellipsis buttons have the same hover and active styles as the normal buttons, but nothing happens when you click them. Would be nice if they don't have hover/active styles so the user doesn't think something should happen.

File: design-library/src/components/BccPagination/BccPagination.css [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        @apply flex flex-row items-center justify-between
+    }
+
+    .bcc-pagination-left {
+        @apply flex flex-row-reverse items-center justify-between
+    }
+
+    .bcc-pagination-center {
+        @apply flex flex-row items-center justify-evenly
+    }
```
Comment (laurensgroeneveld) [inline design-library/src/components/BccPagination/BccPagination.css]: There's some duplication going on here. The `bcc-pagination` base class is not used at all. I would suggest to scrap the `right` class (since it is the default, move it to the base `bcc-pagination` class) and instead implement only the `left`/'center` classes with only the needed modifier classes (they will overwrite anything on the base class).

File: design-library/src/components/BccPagination/BccPagination.css [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        @apply px-3.5 bg-button-secondary text-secondary;
+    }
+
+    .bcc-pagination-button-left {
+        @apply rounded-l-lg px-3.5 text-interactive;
+    }
+
+    .bcc-pagination-button-right {
+        @apply rounded-r-lg px-3.5 text-interactive;
+    }
```
Comment (laurensgroeneveld) [inline design-library/src/components/BccPagination/BccPagination.css]: Same comment on consolidating these in one `bcc-pagination-button` (or just call them `bcc-pagination-arrow`?) and then a modifier class that only changes the rounding.

File: design-library/src/components/BccPagination/BccPagination.vue [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+import { ArrowBackIosIcon } from "@bcc-code/icons-vue";
+import { ArrowForwardIosIcon } from "@bcc-code/icons-vue";
+import { toRefs, ref, computed, watchEffect } from "vue";
+import type { Item } from "../BccTable/BccTable.vue";
+import BccSelect from "../BccSelect/BccSelect.vue";
+import BccFormLabel from "../BccFormLabel/BccFormLabel.vue";
+
+type Props = {
+  tableItems: Item[];
+  rowsPerPageArray?: number[];
```
Comment (laurensgroeneveld) [inline design-library/src/components/BccPagination/BccPagination.vue]: Two small suggestions:
- This prop is optional but it doesn't have a default. I suggest to add one, like 5/10/25/50 or something.
- The name feels a little unclear to me, suggestion: `rowsPerPageOptions`.

File: design-library/src/components/BccPagination/BccPagination.vue
Comment (laurensgroeneveld) [inline design-library/src/components/BccPagination/BccPagination.vue:1]: If I set `displayRowsPerPage` to `false`, the alignment is inverted, so `right` alignment renders the page numbers on the left and vice versa:

image

File: design-library/src/components/BccPagination/BccPagination.vue [hunk 7]
```diff
@@ -0,0 +1,143 @@
+<script setup lang="ts">
+import { ArrowBackIosIcon } from "@bcc-code/icons-vue";
+import { ArrowForwardIosIcon } from "@bcc-code/icons-vue";
+import { toRefs, ref, computed, watchEffect } from "vue";
+import type { Item } from "../BccTable/BccTable.vue";
+import BccSelect from "../BccSelect/BccSelect.vue";
+import BccFormLabel from "../BccFormLabel/BccFormLabel.vue";
+
+type Props = {
+  tableItems: Item[];
```
Comment (laurensgroeneveld) [inline design-library/src/components/BccPagination/BccPagination.vue]: For discussion: I'm not sure if `tableItems` needs to be part of this component. Instead, I feel like we should build pagination into the table component, so that that is the only component that is concerned about the actual data (this doesn't have to happen necessarily in this PR though).

It's still good to have this as a separate component, but instead I feel like it should only concern itself with the actual pagination. That way it can be used for other usecases as well, not just together with our table component. Instead of passing a full `tableItems` array, we could maybe just pass the amount of items so the amount of pages can be correctly calculated?

Curious to hear your opinion on this!

**Review by laurensgroeneveld — CHANGES_REQUESTED** [review]
Thanks a lot, this is an important component currently missing from the library! I have quite a few comments but most of them are pretty small code quality/design things. Functionality wise the component is already very good 👍 Let me know if you don't agree with something 😊

File: design-library/src/components/BccPagination/BccPagination.css (hunk 5 above)
Comment (starkris51) [reply design-library/src/components/BccPagination/BccPagination.css]: when I fixed the double border, the right arrow needed different tailwind code and its so little shared code that I don't think its worth having a separate class

File: design-library/src/components/BccPagination/BccPagination.vue (hunk 7 above)
Comment (starkris51) [reply design-library/src/components/BccPagination/BccPagination.vue]: i do agree that it should not only be used for the table, the change I've did is that it supports array with any types.

Comment (starkris51) [reply design-library/src/components/BccPagination/BccPagination.vue]: @laurensgroeneveld  i have made a alternative version of this using Radix vue libary, it uses passing amount of items instead of a array.
https://github.com/bcc-code/bcc-design/pull/301

File: design-library/src/tokens/tailwind/shadows.ts
Comment (laurensgroeneveld) [reply design-library/src/tokens/tailwind/shadows.ts:1]: The file seems to be still here, despite Github saying this comment is "outdated".

File: design-library/src/components/BccPagination/BccPagination.vue (hunk 7 above)
Comment (laurensgroeneveld) [reply design-library/src/components/BccPagination/BccPagination.vue]: > @laurensgroeneveld i have made a alternative version of this using Radix vue libary, it uses passing amount of items instead of a array.

Yes, that's what I meant here as well. I'll take a look.

> i do agree that it should not only be used for the table, the change I've did is that it supports array with any types.

I feel like the pagination component should not need to be aware of the actual items at all, but instead just the total amount.

File: design-library/src/components/BccPagination/BccPagination.vue [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+</script>
+
+<template>
+  <div
+    :class="{
+      'bcc-pagination-left':
+        (align === 'left' && displayRowsPerPage === true) ||
+        (align === 'right' && displayRowsPerPage === false),
+      'bcc-pagination-center': align === 'center',
+      'bcc-pagination': align === 'right',
```
Comment (laurensgroeneveld) [inline design-library/src/components/BccPagination/BccPagination.vue]: Since the underlying CSS is mostly the same, I think it makes more sense to have `bcc-pagination` as a default class and then just adding the `-left` and `-center` variants in addition when applicable. That's how it's done in most other components as well, for example the secondary button is `bcc-button bcc-button-secondary`, with the last class just adding or overriding classes that are different from the default.

File: design-library/src/components/BccPagination/BccPagination.vue [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+
+const changePage = (value: number) => {
+  currentPage.value += value;
+  emit("update:currentPage", currentPage.value);
+};
+
+watchEffect(() => {
+  emit("update:paginatedItems", paginatedRows.value);
+  emit("update:totalPages", totalPages.value);
+});
```
Comment (laurensgroeneveld) [inline design-library/src/components/BccPagination/BccPagination.vue]: Nice, I should use `watchEffect` more instead of plain `watch`, it works nicer for instances like this 👍

**Review by laurensgroeneveld — COMMENTED** [review]
The CSS is much more consolidated, looks really good 👍 Thanks for picking up all the comment, sorry for late re-review 😅 Just have three small comments (two I added as a comment to previous open threads), other than that it looks good. If you can just remove that `shadows` file I'll approve it 😊

**Comment by starkris51** [comment]
@laurensgroeneveld i'm wondering about if the Table in pagination storybook should be kept or removed like with the Radix Vue version where i only show the component itself.

**Comment by laurensgroeneveld** [comment]
> @laurensgroeneveld i'm wondering about if the Table in pagination storybook should be kept or removed like with the Radix Vue version where i only show the component itself.

I think that's fine, then when we add pagination support to the Table component itself it can instead be shown there.

I didn't add a comment on it before, but you could even add a few stories showing some different options at a glance, like if you hide the page selector, alignment etc.

**Comment by starkris51** [comment]
> > @laurensgroeneveld i'm wondering about if the Table in pagination storybook should be kept or removed like with the Radix Vue version where i only show the component itself.
> 
> I think that's fine, then when we add pagination support to the Table component itself it can instead be shown there.
> 
> I didn't add a comment on it before, but you could even add a few stories showing some different options at a glance, like if you hide the page selector, alignment etc.

I've moved the table as a story instead.

**Review by laurensgroeneveld — APPROVED** [review]
Thanks a lot 👍
